Feedback control update

PoloContreras opened the pull request:
Including feedback control law to guarantee asymptotic convergence to goal location, and including multiplier values "k1" and "k2" to calibrate steering maneuvers.

PoloContreras committed:
Feedback control update

Including feedback control law to guarantee asymptotic convergence to goal location, and including multiplier values "k1" and "k2" to calibrate steering maneuvers.

lowjunen commented on src/bridge_iroad/src/iroad_closedloop.cpp:
```diff
@@ -91,32 +102,46 @@ void Teleop_IRoad::udp_cb(const ros::TimerEvent& event) {
+  /*steer = rad2deg(acos(r_tb.dot(r_f)/d_t)); //debug: steering setting is equal to angle between vehicle heading and goal direction
 
-  steer = min(steer, 30.0);
+  steer = min(steer, max_steer);
```
wouldn't this line be redundant given your definition later in the code?

PoloContreras replied:
The line was only commented out and not deleted outright in case we wanted to restore the prior definition of the value for debugging purposes, but wouldn't run (and is superseded by the later definition that isn't commented out).

lowjunen commented on src/bridge_iroad/src/iroad_closedloop.cpp:
```diff
@@ -91,32 +102,46 @@ void Teleop_IRoad::udp_cb(const ros::TimerEvent& event) {
 
-  steer = min(steer, 30.0);
+  steer = min(steer, max_steer);
+  steer /= max_steer;
```
this line too.

PoloContreras replied:
See above.

lowjunen commented on src/bridge_iroad/src/iroad_closedloop.cpp:
```diff
@@ -91,32 +102,46 @@ void Teleop_IRoad::udp_cb(const ros::TimerEvent& event) {
+  }
+  //double delta; //angle between vehicle heading and desired vehicle heading at target (in radians)
+
+  steer = rad2deg(atan2(whlbase*(ctrl_k2*alpha + ctrl_k1*sin(alpha)*cos(alpha)),v_const))/max_steer; //calculation of steering without regard for desired heading at target
```
split this into smaller equations. i can't quite understand what you're trying to do here.

PoloContreras committed:
Multi-step steering calculation

Divided calculation of closed-loop steering angle (based on feedback control law) into multiple steps for improved readability.

PoloContreras commented on src/bridge_iroad/src/iroad_closedloop.cpp:
```diff
@@ -91,32 +102,49 @@ void Teleop_IRoad::udp_cb(const ros::TimerEvent& event) {
+  //double omega = ctrl_k2*alpha + ctrl_k1*sin(alpha)*cos(alpha)*(alpha + ctrl_k3*delta)/alpha; //instantaneous angular velocity required by feedback control law, if goal position includes a target heading
+
+  steer = rad2deg(atan2(whlbase*omega,v_const)); //conversion of required angular velocity of vehicle to steering angle that will result in that angular velocity
+  steer /= max_steer; //rescale steer value to be a proportion of maximum extent of steering mechanism
```
Split the calculation of the steering angle from the feedback control law into several steps and added some more comments.

PoloContreras reviewed (commented):
Addressed comments & modified code to improve readability.

PoloContreras committed:
Simplified feedback

Simplified feedback rule when not requiring a specific heading (with respect to world frame) as part of the target waypoint.

PoloContreras committed:
Added clarification

Added clarifying comments, functionality is the same
